partiql-logical-planner/src/name_resolver.rs

```diff
@@ -140,8 +152,8 @@ impl NameResolver {
     }
 
     #[inline]
-    fn exit_lateral(&mut self) -> Vec<ast::NodeId> {
-        self.lateral_stack.pop().expect("lateral level")
+    fn exit_lateral(&mut self) -> Option<Vec<ast::NodeId>> {
```

Comment:
I think it might be cleaner to construct error messages here than at each call site.

E.g., 
```rust
    fn exit_lateral(&mut self) -> Result<Vec<ast::NodeId>, LowerError> {
        self.lateral_stack
            .pop()
            .ok_or_else(|| LowerError::IllegalState("Expected non-empty lateral stack".to_string()))
    }
```
Then each call site could be simplified to:
```rust
        if let Err(e) = self.exit_lateral() {
            self.errors.push(e);
            return Recurse::Stop;
        }
```

Similar for the other state stack methods.